qiime2/sdk/proxy.py

```diff
@@ -0,0 +1,193 @@
+            return f'<visualization: Unknown Type {object.__repr__(self)}>'
+        else:
+            return \
+                f'<visualization: ' \
+                f'{self._signature_[self._selector_].qiime_type}>'
+
+
+class ProxyCollection(Proxy):
+    def __init__(self, future, selector, signature=None):
+        self._future_ = future
+        self._selector_ = selector
+        self._signature_ = signature
+
+    def keys(self):
+        return self.result().collection.keys()
+
+    def values(self):
+        return self.result().collection.values()
+
+    def items(self):
+        return self.result().collection.items()
+
+    def get_element(self, results):
+        """Get the result we want off of the future we have
+        """
+        return getattr(results, self._selector_)
+
+    def result(self):
+        return self.get_element(self._future_.result())
+
+    def save(self, directory):
+        """Blocks then calls save on the result.
+        """
+        return self.result().save(directory)
+
+
+class ProxyResults(Proxy):
+    """This represents future results that are being returned by a Parsl app
+    """
+    def __init__(self, future, signature):
+        """We have the future results and the outputs portion of the signature
+        of the action creating the results
+        """
+        self._future_ = future
+        self._signature_ = signature
+
+    def __iter__(self):
+        """Give us a ProxyArtifact for each result in the future
+        """
+        for s in self._signature_:
+            yield self._create_proxy(s)
+
+    def __getattr__(self, attr):
+        """Get a particular ProxyArtifact out of the future
+        """
+        return self._create_proxy(attr)
+
+    def __getitem__(self, index):
+        return self._create_proxy(list(self._signature_.keys())[index])
+
+    def __repr__(self):
+        lines = []
+        lines.append('%s (name = value)' % self.__class__.__name__)
+        lines.append('')
+
+        max_len = -1
+        for field in self._signature_:
+            if len(field) > max_len:
+                max_len = len(field)
+
+        for field, value in zip(self._signature_, self):
+            field_padding = ' ' * (max_len - len(field))
+            lines.append('%s%s = %r' % (field, field_padding, value))
+
+        max_len = -1
+        for line in lines:
+            if len(line) > max_len:
+                max_len = len(line)
+        lines[1] = '-' * max_len
+
+        return '\n'.join(lines)
+
+    def _create_proxy(self, selector):
+        qiime_type = self._signature_[selector].qiime_type
+
+        if is_collection_type(qiime_type):
+            return ProxyCollection(self._future_, selector, self._signature_)
+        elif is_visualization_type(qiime_type):
+            return ProxyVisualization(
+                self._future_, selector, self._signature_)
+
+        return ProxyArtifact(self._future_, selector, self._signature_)
+
+    def result(self):
```

Comment:
This probably needs to be `_result()` to avoid namespace collisions with `__getattr__`/registered outputs